crates/oxc_linter/src/rules/oxc/no_accumulating_spread.rs

```diff
@@ -111,63 +126,142 @@ impl Rule for NoAccumulatingSpread {
         let Some(declaration) = ctx.semantic().nodes().parent_node(declaration_id) else {
             return;
         };
-        let AstKind::FormalParameters(params) = declaration.kind() else {
-            return;
-        };
 
-        // We're only looking for the first parameter, since that's where acc is.
-        // Skip non-parameter or non-first-parameter declarations.
-        let first_param_symbol_id =
-            params.items.first().and_then(|item| get_identifier_symbol_id(&item.pattern.kind));
-        if !first_param_symbol_id.is_some_and(|id| id == referenced_symbol_id) {
-            return;
-        }
+        check_reduce_usage(declaration, referenced_symbol_id, spread.span, ctx);
+        check_loop_usage(
+            declaration,
+            ctx.semantic().nodes().get_node(declaration_id),
+            referenced_symbol_id,
+            node.id(),
+            spread.span,
+            ctx,
+        );
+    }
+}
 
-        // invalid number of parameters to reduce callback
-        let params_count = params.parameters_count();
-        if params_count != 2 {
+fn check_reduce_usage<'a>(
+    declaration: &AstNode<'a>,
+    referenced_symbol_id: SymbolId,
+    spread_span: Span,
+    ctx: &LintContext<'a>,
+) {
+    let AstKind::FormalParameters(params) = declaration.kind() else {
+        return;
+    };
+
+    // We're only looking for the first parameter, since that's where acc is.
+    // Skip non-parameter or non-first-parameter declarations.
+    let first_param_symbol_id =
+        params.items.first().and_then(|item| get_identifier_symbol_id(&item.pattern.kind));
+    if !first_param_symbol_id.is_some_and(|id| id == referenced_symbol_id) {
+        return;
+    }
+
+    // invalid number of parameters to reduce callback
+    let params_count = params.parameters_count();
+    if params_count != 2 {
+        return;
+    }
+
+    // Check if the declaration resides within a call to reduce()
+    for parent in ctx.nodes().iter_parents(declaration.id()) {
+        if let AstKind::CallExpression(call_expr) = parent.kind() {
+            if is_method_call(call_expr, None, Some(&["reduce", "reduceRight"]), Some(1), Some(2)) {
+                ctx.diagnostic(get_reduce_diagnostic(call_expr, spread_span));
+            }
             return;
         }
+    }
+}
+
+fn check_loop_usage<'a>(
+    declaration_node: &AstNode<'a>,
+    declarator: &AstNode<'a>,
+    referenced_symbol_id: SymbolId,
+    spread_node_id: AstNodeId,
+    spread_span: Span,
+    ctx: &LintContext<'a>,
+) {
+    let AstKind::VariableDeclaration(declaration) = declaration_node.kind() else {
+        return;
+    };
+
+    if !matches!(declaration.kind, VariableDeclarationKind::Let) {
```

Comment:
What about `for (const x of arr)`?

Reply:
so this line is checking that the declaration

```ts
let foo = []; for (let i = 0; i < 10; i++) { foo = [...foo, i]; }
^^^^^^^^
```

here is `let` - if it's not a `let` declaration, then that means that it must never be reassigned hence not a violation of this rule.

this is a good point though. i've added some test cases for `for (const ...` cases

Reply:
maybe it's worth adding some code comments for clarity?